lib/_emerge/actions.py

```diff
@@ -2655,18 +2655,37 @@ def nice(settings):
 		out = portage.output.EOutput()
 		out.eerror(f"Failed to change nice value to {nice_value}")
 		out.eerror("%s\n" % str(e))
+		return
 
-	autogroup_file = Path("/proc/self/autogroup")
-	try:
-		f = autogroup_file.open("r+")
-	except EnvironmentError:
+	sched_autogroup = Path("/proc/sys/kernel/sched_autogroup_enabled")
+	if not sched_autogroup.is_file():
 		# Autogroup scheduling is not enabled on this system.
 		return
 
-	with f:
+	try:
+		# Will be "1" (enabled) or "0" disabled.
+		sched_autogroup_enabled = int(sched_autogroup.read_text())
```

Comment:
The kind of TOCTOU race that I want to be sure we handle would be where the /proc/sys/kernel/sched_autogroup_enabled state changes before this function returns.

Reply:
I think that the problem in my prefix system is that the kernel is just too old (2.6.32). Even when the feature is enabled and I just do `echo 10 > /proc/self/autogroup` as root, it fails.

Reply:
> The kind of TOCTOU race that I want to be sure we handle would be where the /proc/sys/kernel/sched_autogroup_enabled state changes before this function returns.

With this PR, If the value changes then between reading `sched_autogroup_enabled` and writing `/proc/self/autogroup` from 1 to 0, then the write will fail with an error message and the `atexit()` handler won't be installed. I think that is sensible, even though the TOCTOU still exists, but "only" results in one error message, and not two, as with the current master (one for the write exception and one from the atexit handler write throwing).

I am not sure how this TOCTOU can be prevented in this case, but maybe I am missing something.

Reply:
> This check can actually be removed,

I tend towards keeping the check, as it prevents an error message, due the exception thrown by `write(/proc/self/autogroup)`, in case `sched_autogroup_enabled` is 0.

Reply:
On a different system with newer kernel (3.10), no errors occur, whether the feature is enabled or not. The value just doesn't change if the feature is disabled. Only with kernel 2.6.32 it fails no matter what I do.